keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,351 @@
+
+#### Data Compatibility Risks during Feature Value Change
+
+There are some features that would need to touch data in the apply/storage layer. Disabling them might result in orphaned data or unrecognizable data that could be breaking.
+How do we handle the data change? 
+1. if the etcd server knows this feature, it should be able to handle the data when the feature is on or off. How it handles the data at WAL entry `i` is determined by the cluster feature flag set from raft before WAL entry `i`. 
+1. if the etcd server does not know this feature, i.e. its version is not compatible with this feature, it means the data is not compatible with the storage version. The storage version migration should handle cleaning up older or newly introduced data.
+
+Overall, when developers introduce new data with a feature, they should be careful to include data cleaning in `schema.UnsafeMigrate`. But we prefer not to have a mechanism to disallow such features to be turned off after they are turned on because new features could be buggy, we need to keep the option to be able to turn it off without bring down the whole cluster.
+
+#### Feature Implementation Change Risks
+
+There are cases when a developer wants to change how an existing feature works. This would make things complicated in a mixed version cluster. One member could have the old implementation while another member uses the latest implementation. How can we make sure the cluster works consistently?
+
+Ideally, the developer should introduced a new feature if they want to do major changes. If that is not desirable, similar to [how Kubernetes handles feature gating changes for compatibility version](https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/4330-compatibility-versions#feature-gating-changes), we need to keep the implementation history for at least 2 minor versions (allowed version skew) in the code base if the feature change affects data in anyway, like
+```go
+if s.FeatureEnabled(FeatureA) && s.ClusterVersion() <= "3.6" {implementation 1}
+if s.FeatureEnabled(FeatureA) && s.ClusterVersion() >= "3.7" {implementation 2}
+```
+This way the cluster would work consistently because there is a single ClusterVersion across the whole cluster.
+
+## Design Details
+
+### APIs
+
+Cluster feature gate would share the same APIs as the server feature gate. 
+
+A feature can be registered as server level feature or cluster level feature, but not both. So for all the feature enablement query apis, 
+* if the feature is a sever level feature, the result would be if the feature is enabled for that server.
+* if the feature is a cluster level feature, the result would be if the feature is enabled for the whole cluster.
+
+### Feature Gate
+
+A feature can be registered as server level feature or cluster level feature, but not both.
```

Comment:
Is there a possibility of converting or migrating a `server level feature` to `cluster level feature` in later releases?

Reply:
Migration would mean we remove it from server level feature gate, which means a breaking change in cli.
So default answer should be no. We might still consider to have a process for migrating by deprecating a server feature and creating a similarly named cluster level feature, but we don't need to discuss it the KEP.